api/traces/v1/http.go

```diff
@@ -0,0 +1,291 @@
+			DialContext: (&net.Dialer{
+				Timeout: ReadTimeout,
+			}).DialContext,
+		}
+
+		proxyRead = &httputil.ReverseProxy{
+			Director:     middlewares,
+			ErrorLog:     proxy.Logger(c.logger),
+			Transport:    otelhttp.NewTransport(t),
+			ErrorHandler: func(rw http.ResponseWriter, r *http.Request, e error) {},
+		}
+	}
+	r.Group(func(r chi.Router) {
+		r.Use(c.readMiddlewares...)
+		r.Handle("/*", c.instrument.NewHandler(
+			prometheus.Labels{"group": "tracesv1api", "handler": "api"},
+			proxyRead))
+	})
+
+	return r
+}
+
+// NewUIStaticHandler creates a trace handler for Jaeger UI static assets (.js and .css)
+func NewUIStaticHandler(read *url.URL, opts ...HandlerOption) http.Handler {
+	c := &handlerConfiguration{
+		logger:     log.NewNopLogger(),
+		registry:   prometheus.NewRegistry(),
+		instrument: nopInstrumentHandler{},
+	}
+
+	for _, o := range opts {
+		o(c)
+	}
+
+	r := chi.NewRouter()
+
+	var proxyRead http.Handler
+	{
+		level.Debug(c.logger).Log("msg", "Configuring upstream Jaeger UI static", "ui", read)
+		middlewares := proxy.Middlewares(
+			proxy.MiddlewareSetUpstream(read),
+			proxy.MiddlewareSetPrefixHeader(),
+			proxy.MiddlewareLogger(c.logger),
+			proxy.MiddlewareMetrics(c.registry, prometheus.Labels{"proxy": "tracesv1-ui-static"}),
+		)
+
+		t := &http.Transport{
+			DialContext: (&net.Dialer{
+				Timeout: ReadTimeout,
+			}).DialContext,
+		}
+
+		proxyRead = &httputil.ReverseProxy{
+			Director:  middlewares,
+			ErrorLog:  proxy.Logger(c.logger),
+			Transport: otelhttp.NewTransport(t),
+		}
+	}
+	r.Group(func(r chi.Router) {
+		r.Use(c.readMiddlewares...)
+		r.Handle("/*", c.instrument.NewHandler(
+			prometheus.Labels{"group": "tracesv1ui", "handler": "search"},
+			proxyRead))
+	})
+
+	return r
+}
+
+// NewUIHandler creates a trace handler for Jaeger UI that is able to re-write
+// HTML to change the <base> attribute so that it works with the Observatorium-style
+// "/api/v1/traces/{tenant}/" URLs.
+func NewUIHandler(read *url.URL, opts ...HandlerOption) http.Handler {
+	c := &handlerConfiguration{
+		logger:     log.NewNopLogger(),
+		registry:   prometheus.NewRegistry(),
+		instrument: nopInstrumentHandler{},
+	}
+
+	for _, o := range opts {
+		o(c)
+	}
+
+	r := chi.NewRouter()
+
+	var proxyRead http.Handler
+	{
+		level.Debug(c.logger).Log("msg", "Configuring upstream Jaeger UI", "ui", read)
+		middlewares := proxy.Middlewares(
+			proxy.MiddlewareSetUpstream(read),
+			proxy.MiddlewareSetPrefixHeader(),
+			proxy.MiddlewareLogger(c.logger),
+			proxy.MiddlewareMetrics(c.registry, prometheus.Labels{"proxy": "tracesv1-ui"}),
+		)
+
+		t := &http.Transport{
+			DialContext: (&net.Dialer{
+				Timeout: ReadTimeout,
+			}).DialContext,
+		}
+
+		proxyRead = &httputil.ReverseProxy{
+			Director:  middlewares,
+			ErrorLog:  proxy.Logger(c.logger),
+			Transport: otelhttp.NewTransport(t),
+
+			// This is the key piece, it changes <base href=> tags
+			ModifyResponse: jaegerUIResponseModifier,
```

Comment:
I wonder what happens here in case `jaegerUIResponseModifier` returns an error?
should `ErrorHandler` be implemented as well https://go.dev/src/net/http/httputil/reverseproxy.go ?